### PR TITLE
Offchain Workers: Example Pallet

### PR DESCRIPTION
**Pull request by tomusdrw**
This PR introduces an example pallet, which showcases simple usage of offchain workers. I've tried to focus on documenting all code patterns and some possible edge-cases and write some basic tests for the pallet. It obviously does not cover all cases (like signed payload inside an unsigned transaction), but should be a good enough example to start with. I believe that most complicated examples should become part of the `recipes` repository.

I've decided to make it separate from the `pallet-example` to avoid polluting the module with OCW-specific stuff. I think it's easier to digest both in isolation.

Since I'm not a great writer, I'd appreciate all feedback on the comments and ideas how to rephrase them. Also happy to accept direct commits to that branch to extend the documentation.

There are two significant changes to the substrate codebase, namely:
1. We initialize the logger in `executive`. - Since our core pallets initialize it anyway, it doesn't make much sense to repeat that line in every OCW.
2. Re-export `sp_core::offchain` from `sp_runtime::offchain`. The split there is pretty much arbitrary (the latter contains some traits that are needed by the client as well), so I thought it's easier to just have it in a single place.

### CODE REVIEW

File: frame/example-offchain-worker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+	traits::{Zero, UniqueSaturatedFrom},
+	transaction_validity::{InvalidTransaction, ValidTransaction, TransactionValidity},
+};
+
+#[cfg(test)]
+mod tests;
+
+/// Defines application identifier for crypto keys of this module.
+///
+/// Every module that deals with signatures needs to declare it's unique identifier for
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
/// Every module that deals with signatures needs to declare its unique identifier for
```

File: frame/example-offchain-worker/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+/// it's crypto keys.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
/// its crypto keys.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+/// When offchain worker is signing transactions it's going to request keys with below
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
/// When offchain worker is signing transactions it's going to request keys of type
```

File: frame/example-offchain-worker/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+/// `KeyTypeId` from the keystore and use the ones it founds to sign the transaction.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
/// `KeyTypeId` from the keystore and use the ones it finds to sign the transaction.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+	/// The type to sign and submit transactions.
+	type SubmitSignedTransaction:
+		offchain::SubmitSignedTransaction<Self, <Self as Trait>::Call>;
+	/// The type to submit unsigned transactions.
+	type SubmitUnsignedTransaction:
+		offchain::SubmitUnsignedTransaction<Self, <Self as Trait>::Call>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// The overarching event type.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
	/// The overarching dispatch call type.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+	type Call: From<Call<Self>>;
+}
+
+/// Number of blocks of cooldown after unsigned transaction is included.
+///
+/// This ensure that we only accept unsigned transaction once, every `UNSIGNED_INTERVAL` blocks.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
/// This ensures that we only accept unsigned transactions once, every `UNSIGNED_INTERVAL` blocks.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+		/// we only allow one transaction every UNSIGNED_INTERVAL blocks.
+		/// This storage entry defines when new transaction is going to be accepted.
+		NextUnsignedAt get(fn next_unsigned_at): T::BlockNumber;
+	}
+}
+
+decl_event!(
+	/// Events generated by the module.
+	pub enum Event<T> where AccountId = <T as frame_system::Trait>::AccountId {
+		/// Event generated when new price is accepted to contirbute to the average.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// Event generated when new price is accepted to contribute to the average.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+decl_module! {
+	/// A public part of the pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Submit new price to the list.
+		///
+		/// This method is a public function of the module and can be called from within
+		/// a transaction. It appends given `price` to current list of prices.
+		///	In our example the `offchain worker` will create, sign & submit a transaction that
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// In our example the `offchain worker` will create, sign & submit a transaction that
```

File: frame/example-offchain-worker/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+		///	calls this function passing the price.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// calls this function passing the price.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+		///
+		/// The transaction needs to be signed (see `ensure_signed`) check, so that the caller
+		/// pays fee to execute it.
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// pays a fee to execute it.
```

File: frame/example-offchain-worker/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+		/// This makes sure that it's not easy (or rather cheap) to attack the chain by submitting
+		/// excesive transactions, but note that it doesn't ensure the price oracle is actually
+		/// working and receives (and provides) meaningful data.
+		/// This example is not focused on correctness of the oracle itself, but rather it's
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// This example is not focused on correctness of the oracle itself, but rather its
```

File: frame/example-offchain-worker/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+		/// This function can be called only once every `UNSIGNED_INTERVAL` blocks.
+		/// Transactions that call that function are de-duplicated on the pool level
+		/// via `validate_unsigned` implementation and also are rendered invalid if
+		/// the function has already been called in current "session".
+		///
+		/// It's important to specify `weight` for unsigned calls as well, because even though
+		/// they don't charge fees, we still don't want a single block to contain unlimited
+		/// number of such transactions.
+		///
+		/// This example is not focused on correctness of the oracle itself, but rather it's
```
Comment (gavofyork) [inline frame/example-offchain-worker/src/lib.rs]:
```suggestion
		/// This example is not focused on correctness of the oracle itself, but rather its
```

**Comment by apopiak** [comment]
> I think this should indeed be in the 2.0 release. Just needs to have the spelling nits fixed + and my personal grumble that our preferred line-width repo-wide is 100 chars, not 60 or 80 :D but that might be a story for a different day.

I agree with Joshy's comment that it's odd to have this non-functional example in frame and that it's a better fit for the recipes or somewhere else in the repo (some `pallet-examples` directory or something).

**Comment by tomusdrw** [comment]
> I agree with Joshy's comment that it's odd to have this non-functional example in frame and that it's a better fit for the recipes or somewhere else in the repo (some pallet-examples directory or something).

I think the main premise to have `pallet-example` and `node-template` (and even `node` or `frame`) within the same repo is to make sure they are properly updated along the fast development cycle we have.
I feel that `pallet-example-offchain-worker` is in the same vein now, but I'm happy to revist that after we stabilise 2.0 and extract `frame` to it's own repo.

**Comment by gnunicorn** [comment]
needs merge.

**Pull request merged**